File: lintcode/partition-list.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/partition-list/

from linked_list import *

"""
Definition of ListNode
class ListNode(object):
    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: The first node of linked list.
    @param x: an integer
    @return: a ListNode
    """
    def partition(self, head, x):
        # write your code here
        if not head or not head.next:
            return head

        small = None
        small_cursor = None
        large = None
        large_cursor = None

        while head:
            if head.val < x:
                temp = head.next
                if not small:
                    small = head
                    small_cursor = small
                    small_cursor.next = None
                else:
                    small_cursor.next = head
                    small_cursor = small_cursor.next
                    small_cursor.next = None
                head = temp
            else:
                temp = head.next
                if not large:
                    large = head
                    large_cursor = large
                    large_cursor.next = None
                else:
                    large_cursor.next = head
                    large_cursor = large_cursor.next
                    large_cursor.next = None
                head = temp

        if not small:
            return large
        if not large:
            return small
        if small and large:
            small_cursor.next = large
            return small
        return None

if __name__ == '__main__':
    # head = create_linked_list([1,4,3,2,5,2])
    head = create_linked_list([3,3,1,2,4])
    s = Solution()
    new_head = s.partition(head, 3)
    print_linked_list(new_head)

File: leetcode-2/partition-list.cc
// Question: https://leetcode.com/problems/partition-list/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode dummy1(0), dummy2(0);
        auto node = head;
        auto cursor1 = &dummy1;
        auto cursor2 = &dummy2;
        while (node) {
            auto next = node->next;
            if (node->val < x) {
                node->next = cursor1->next;
                cursor1->next = node;
                cursor1 = node;
            } else {
                node->next = cursor2->next;
                cursor2->next = node;
                cursor2 = node;
            }
            node = next;
        }
        cursor1->next = dummy2.next;
        return dummy1.next;
    }
};
